operators/src/processing/new_expression/query_processor.rs

```diff
@@ -0,0 +1,401 @@
+use std::{marker::PhantomData, sync::Arc};
+
+use async_trait::async_trait;
+use futures::{stream::BoxStream, try_join, StreamExt, TryStreamExt};
+use geoengine_datatypes::{
+    primitives::{SpatialPartition2D, TimeInterval},
+    raster::{
+        GeoTransform, Grid2D, GridIdx2D, GridShape2D, GridShapeAccess, NoDataValue, Pixel,
+        RasterTile2D,
+    },
+};
+use num_traits::AsPrimitive;
+use rayon::iter::{IntoParallelIterator, IntoParallelRefIterator, ParallelIterator};
+
+use crate::{
+    engine::{BoxRasterQueryProcessor, QueryContext, QueryProcessor, RasterQueryRectangle},
+    util::{stream_zip::StreamTupleZip, Result},
+};
+
+use super::compiled::LinkedExpression;
+
+pub struct ExpressionQueryProcessor<TO, Sources>
+where
+    TO: Pixel,
+{
+    pub sources: Sources,
+    pub phantom_data: PhantomData<TO>,
+    pub program: Arc<LinkedExpression>,
+    pub no_data_value: TO,
+    pub map_no_data: bool,
+}
+
+impl<TO, Sources> ExpressionQueryProcessor<TO, Sources>
+where
+    TO: Pixel,
+{
+    pub fn new(
+        program: LinkedExpression,
+        sources: Sources,
+        no_data_value: TO,
+        map_no_data: bool,
+    ) -> Self {
+        Self {
+            sources,
+            program: Arc::new(program),
+            phantom_data: PhantomData::default(),
+            no_data_value,
+            map_no_data,
+        }
+    }
+}
+
+#[async_trait]
+impl<'a, TO, Tuple> QueryProcessor for ExpressionQueryProcessor<TO, Tuple>
+where
+    TO: Pixel,
+    Tuple: ExpressionTupleProcessor<TO>,
+{
+    type Output = RasterTile2D<TO>;
+    type SpatialBounds = SpatialPartition2D;
+
+    async fn query<'b>(
+        &'b self,
+        query: RasterQueryRectangle,
+        ctx: &'b dyn QueryContext,
+    ) -> Result<BoxStream<'b, Result<Self::Output>>> {
+        let stream = self
+            .sources
+            .queries(query, ctx)
+            .await?
+            .and_then(move |rasters| async move {
+                if Tuple::all_empty(&rasters) {
+                    return Ok(Tuple::empty_raster(&rasters));
+                }
+
+                let (out_time, out_tile_position, out_global_geo_transform, output_grid_shape) =
+                    Tuple::metadata(&rasters);
+
+                let out_no_data = self.no_data_value;
+
+                let thread_pool = ctx.thread_pool().clone();
+                let program = self.program.clone();
+                let map_no_data = self.map_no_data;
+
+                let data = tokio::task::spawn_blocking(move || {
+                    thread_pool.install(move || {
+                        Tuple::compute_expression(rasters, &program, map_no_data, out_no_data)
+                    })
+                })
+                .await??;
+
+                let out =
+                    Grid2D::<TO>::new(output_grid_shape, data, Some(self.no_data_value))?.into();
+
+                Ok(RasterTile2D::new(
```

Comment:
where do you transform the pixels into the desired output data type?

Reply:
`data` has to be of type `TO` because of the `Grid2D::<TO>` bound here.

For instance, in line 199, you can spot `TO::from_(result)`.